File: Assets/Game/Script/Framework/UI/UIView/UIBaseView.cs
using System;
using ZJYFrameWork.UISerializable.UIInitView;
using UnityEngine;
using UnityEngine.UI;

namespace ZJYFrameWork.UISerializable
{
    public abstract class UIBaseView<PanelView> where PanelView : UIViewInterface, new()
    {
        /// <summary>
        /// UI实例化时调用 OnInit
        /// </summary>
        public abstract void OnInit();

        /// <summary>
        /// 当物体打开时调用
        /// </summary>
        public abstract void OnShow();

        /// <summary>
        /// 当Canvas CanvasGroup GraphicRaycaster 设置为不激活时调用
        /// </summary>
        public virtual void OnHide()
        {
            UnityEngine.Debug.Log("隐藏");
            SelfUIView.OnClose();
        }

        [Obsolete("请调用OnHide方法")]
        public virtual void OnClickDestroy()
        {
        }

        public void ReUse()
        {
            SelfUIView.OnShow();
            OnShow();
        }

        /// <summary>
        /// 物体是否隐藏
        /// </summary>
        public bool IsActive
        {
            get
            {
                if (SelfUIView != null)
                {
                    if (SelfUIView.GetSelfObjCanvasGroup)
                    {
                        var isActive = SelfUIView.GetSelfObjCanvasGroup.alpha > 0;
                        return isActive;
                    }
                    else
                    {
#if UNITY_EDITOR
                        // Debug.LogError(
                        //     $"UIView组件 缺少{typeof(Canvas)},缺少{typeof(GraphicRaycaster)},缺少{typeof(CanvasGroup)},");
#endif
                    }

                    return false;
                }

                return false;
            }
        }

        /// <summary>
        /// 每个预制体都必须需要的UIview
        /// </summary>
        private UIView SelfUIView;

        public UIView GetSelfUIView => SelfUIView;

        /// <summary>
        /// 为ui显示面板 所创建脚本存放所有要使用的脚本 UI组件全部通过这里去调度
        /// </summary>
        public PanelView viewPanel;


        public void SetUIView(UIView view)
        {
            SelfUIView = view;
            viewPanel = new PanelView();
            viewPanel.Init(SelfUIView);
        }
    }
}

File: DemoMMOUnity/Lua/Common/UI/UIBaseView.lua
local UIBaseView = BaseClass()
local viewPanel={}
function UIBaseView:Create(uiPanelView)
    local uiView = UIBaseView:New();

    -- uiView.SetUIView(View,uiPanelView)
    self.viewPanel=uiPanelView;
    return uiView

end
function UIBaseView:ReUse()
    self.SelfUIView:OnShow()
    UIBaseView:OnShow()
end

function UIBaseView:OnShow()
end
-- function UIBaseView:SetUIView(view)
--     print("初始化 set"..view)
--     local uiView = view:GetComponent("UIView")
--     self.SelfUIView = uiView
--     print("UIBaseView:SetUIView Init"..self.SelfUIView)
--     self.viewPanel.Init(view)
-- end
--- 隐藏 UI
function UIBaseView:OnHide()
    self.SelfUIView.OnClose()
end
UIBaseView.viewPanel=viewPanel
return UIBaseView